common/unsafe/src/main/java/org/apache/spark/unsafe/memory/UnsafeMemoryAllocator.java

```diff
@@ -36,22 +42,35 @@ public MemoryBlock allocate(long size) throws OutOfMemoryError {
 
   @Override
   public void free(MemoryBlock memory) {
-    assert (memory.obj == null) :
+    assert(memory instanceof OffHeapMemoryBlock);
+    if (memory == OffHeapMemoryBlock.NULL) return;
+    assert (memory.getBaseObject() == null) :
```

Comment:
Will this assert be redundant since ```memory```  has already been verified to be a ```OffHeapMemoryBlock ```. And I think, since we introduce multiple types of ```MemoryBlock```, we can leave the tips after the ```assert(memory instanceof OffHeapMemoryBlock)```, like ```"UnsafeMemoryAllocator can only free OffHeapMemoryBlock."``` WDYT?

Reply:
Good catch

Reply:
> its object may not be``` null```

Hi, @kiszk. Actually, I think, here, object must be ```null```. Because I only see one constructor of ```OffHeapMemoryBlock ```:
```
public OffHeapMemoryBlock(long address, long size) {
   super(null, address, size);
 }
```
WDYT?

Reply:
I see. To remove this assertion would not change the behavior.